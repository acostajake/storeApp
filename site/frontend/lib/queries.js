import gql from 'graphql-tag';

import { perPage } from '../config';

export const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$title: String!
		$desc: String!
		$image: String
		$largeImg: String
		$price: Int!
	) {
		createItem(
			title: $title
			desc: $desc
			image: $image
			largeImg: $largeImg
			price: $price
		) {
			id
		}
	}
`;

export const DELETE_ITEM_MUTATION = gql`
	mutation DELETE_ITEM_MUTATION($id: ID!) {
		deleteItem(id: $id) {
			id
		}
	}
`;

export const UPDATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$id: ID!
		$title: String
		$desc: String
		$price: Int
	) {
		updateItem(id: $id, title: $title, desc: $desc, price: $price) {
			id
			title
			desc
			price
		}
	}
`;

export const GET_ALL_ITEMS = gql`
	query GET_ALL_ITEMS($skip: Int = 0, $first: Int = ${perPage}) {
		items(first: $first, skip: $skip, orderBy: createdAt_DESC) {
			id
			title
			desc
			image
			largeImg
			price
		}
	}
`;

export const GET_ITEM = gql`
	query GET_ITEM($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			desc
			price
		}
	}
`;

export const GET_ITEM_WITH_IMAGE = gql`
	query GET_ITEM_WITH_IMAGE($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			desc
			largeImg
		}
	}
`;

export const PAGINATION_QUERY = gql`
	query PAGINATION_QUERY {
		itemsConnection {
			aggregate {
				count
			}
		}
	}
`;

// Auth
export const RESET_MUTATION = gql`
	mutation RESET_MUTATION(
		$resetToken: String!
		$password: String!
		$confirmPassword: String!
	) {
		resetPassword(
			resetToken: $resetToken
			password: $password
			confirmPassword: $confirmPassword
		) {
			id
			name
			email
		}
	}
`;

export const REQUEST_RESET_MUTATION = gql`
	mutation REQUEST_RESET_MUTATION($email: String!) {
		requestReset(email: $email) {
			message
		}
	}
`;

export const SIGNUP_MUTATION = gql`
	mutation SIGNUP_MUTATION(
		$name: String!
		$email: String!
		$password: String!
	) {
		signup(name: $name, email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGNIN_MUTATION = gql`
	mutation SIGNIN_MUTATION($email: String!, $password: String!) {
		signin(email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGNOUT_MUTATION = gql`
	mutation SIGNOUT_MUTATION {
		signout {
			message
		}
	}
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
	mutation updatePermissions($permissions: [Permission], $userId: ID!) {
		updatePermissions(permissions: $permissions, userId: $userId) {
			id
			permissions
			name
			email
		}
	}
`;

export const CURRENT_USER = gql`
	query {
		me {
			id
			email
			name
			permissions
		}
	}
`;

export const GET_ALL_USERS = gql`
	query {
		users {
			id
			name
			email
			permissions
		}
	}
`;
