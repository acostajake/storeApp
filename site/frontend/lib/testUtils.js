import casual from 'casual';
import {
	CURRENT_USER,
	PAGINATION_QUERY,
	REQUEST_RESET_MUTATION,
} from './queries';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	image: 'dog-small.jpg',
	title: 'dogs are best',
	desc: 'dogs',
	largeImg: 'dog.jpg',
});

const fakeUser = () => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
});

const fakeOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: `${casual.word}.jpg`,
	title: casual.words(),
	price: 4234,
	quantity: 1,
	desc: casual.words(),
});

const fakeOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [fakeOrderItem(), fakeOrderItem()],
	createdAt: '2018-04 - 06T19: 24: 16.000Z',
	user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	item: fakeItem(),
	user: fakeUser(),
	...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key) {
		return this.store[key] || null;
	}

	setItem(key, value) {
		this.store[key] = value.toString();
	}

	removeItem(key) {
		delete this.store[key];
	}
}

const notSignedInMocks = [
	{
		request: { query: CURRENT_USER },
		result: { data: { me: null } },
	},
];

const signedInMocks = [
	{
		request: { query: CURRENT_USER },
		result: { data: { me: fakeUser() } },
	},
];

const signedInWithCartMocks = [
	{
		request: { query: CURRENT_USER },
		result: {
			data: {
				me: {
					...fakeUser(),
					cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
				},
			},
		},
	},
];

const makeMocksFor = (length) => {
	return [
		{
			request: { query: PAGINATION_QUERY },
			result: {
				data: {
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							__typename: 'count',
							count: length,
						},
					},
				},
			},
		},
	];
};

const requestResetMock = [
	{
		request: {
			query: REQUEST_RESET_MUTATION,
			variables: { email: 'test@fakeemail.com' },
		},
		result: {
			data: {
				requestReset: {
					message: 'success',
					__typename: 'Message',
				},
			},
		},
	},
];

export {
	LocalStorageMock,
	fakeItem,
	fakeUser,
	fakeCartItem,
	fakeOrder,
	fakeOrderItem,
	notSignedInMocks,
	requestResetMock,
	signedInMocks,
	signedInWithCartMocks,
	makeMocksFor,
};
