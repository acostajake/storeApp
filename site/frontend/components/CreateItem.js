import Router from 'next/router';
import React, { useState } from 'react';
import { Mutation } from 'react-apollo';

import { CREATE_ITEM_MUTATION } from '../lib/queries';
import Form from './styles/Form';
import Error from './ErrorMessage';

const CreateItem = () => {
	const [item, setItemDetails] = useState({
		title: '',
		desc: '',
		image: '',
		largeImg: '',
		price: '',
	});

	const updateInput = (e) => {
		const {
			target: { name, type, value },
		} = e;
		const val = type === 'number' ? parseFloat(value) : value;
		setItemDetails({ ...item, [name]: val });
	};

	const uploadImage = async (e) => {
		const files = e.target.files;
		const data = new FormData();
		data.append('file', files[0]);
		data.append('upload_preset', 'storeApp');
		const res = await fetch(
			'https://api.cloudinary.com/v1_1/jakefromsnakefarm/image/upload',
			{
				method: 'POST',
				body: data,
			}
		);
		const file = await res.json();
		setItemDetails({
			image: file.secure_url,
			largeImg: file.eager[0].secure_url,
		});
	};

	return (
		<Mutation mutation={CREATE_ITEM_MUTATION} variables={{ ...item }}>
			{(createItem, { loading, error }) => (
				<Form
					onSubmit={async (e) => {
						e.preventDefault();
						const res = await createItem();
						Router.push({
							pathname: '/item',
							query: { id: res.data.createItem.id },
						});
					}}
				>
					<h2>Sell Item</h2>
					<Error error={error} />
					<fieldset disabled={loading} aria-busy={loading}>
						<label htmlFor='image'>
							Add Image
							<input
								type='file'
								id='file'
								name='image'
								placeholder='Add an image'
								onChange={(e) => uploadImage(e)}
							/>
							{item.image && (
								<img width='160px' src={item.image} alt='Image preview' />
							)}
						</label>
						<label htmlFor='title'>
							Title
							<input
								type='text'
								id='title'
								name='title'
								placeholder='Title'
								required
								onChange={(e) => updateInput(e)}
								value={item.title}
							/>
						</label>
						<label htmlFor='price'>
							Price
							<input
								type='number'
								id='price'
								name='price'
								placeholder='Price'
								required
								onChange={(e) => updateInput(e)}
								value={item.price}
							/>
						</label>
						<label htmlFor='description'>
							Description
							<textarea
								type='text'
								id='description'
								name='desc'
								placeholder='Description'
								required
								onChange={(e) => updateInput(e)}
								value={item.desc}
							/>
						</label>
						<button type='submit'>Submit</button>
					</fieldset>
				</Form>
			)}
		</Mutation>
	);
};

export default CreateItem;
