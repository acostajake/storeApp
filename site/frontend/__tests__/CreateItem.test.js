import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';

import CreateItem from '../components/CreateItem';
import { CREATE_ITEM_MUTATION } from '../lib/queries';
import { createItemMock, fakeItem } from '../lib/testUtils';

const dogImage = 'https://dog.com/dog.jpg';

global.fetch = jest.fn().mockResolvedValue({
	json: () => ({
		secure_url: dogImage,
		eager: [{ secure_url: dogImage }],
	}),
});

describe('<CreateItem/>', () => {
	it('matches snapshot', () => {
		const mountedCreateItem = mount(
			<MockedProvider>
				<CreateItem />
			</MockedProvider>
		);
		const form = mountedCreateItem.find('form[data-test="form"]');
		expect(toJSON(form)).toMatchSnapshot();
	});
	it('uploads a file when changed', async () => {
		const wrapper = mount(
			<MockedProvider>
				<CreateItem />
			</MockedProvider>
		);
		const input = wrapper.find('input[type="file"]');
		input.simulate('change', { target: { files: ['fakedog.jpg'] } });
		await wait();
		const component = wrapper.find('CreateItem').instance();
		expect(component.state.image).toEqual(dogImage);
		expect(component.state.largeImg).toEqual(dogImage);
		expect(global.fetch).toHaveBeenCalled();
		global.fetch.mockReset();
	});
	it('can update state by typing', async () => {
		const wrapper = mount(
			<MockedProvider>
				<CreateItem />
			</MockedProvider>
		);
		wrapper
			.find('#title')
			.simulate('change', { target: { value: 'Test', name: 'title' } });
		wrapper
			.find('#price')
			.simulate('change', { target: { value: '50000', name: 'price' } });
		wrapper.find('#description').simulate('change', {
			target: { value: 'This is awesome!', name: 'desc' },
		});
		expect(wrapper.find('CreateItem').instance().state).toMatchObject({
			title: 'Test',
			price: '50000',
			desc: 'This is awesome!',
			image: '',
			largeImg: '',
		});
	});
	it('creates an item on submit', async () => {
		const item = fakeItem();
		const createItemMock = [
			{
				request: {
					query: CREATE_ITEM_MUTATION,
					variables: {
						title: item.title,
						desc: item.desc,
						image: '',
						largeImg: '',
						price: item.price,
					},
				},
				result: {
					data: {
						createItem: {
							...fakeItem,
							id: '12345',
							__typeName: 'Item',
						},
					},
				},
			},
		];
		const wrapper = mount(
			<MockedProvider mocks={createItemMock}>
				<CreateItem />
			</MockedProvider>
		);
		wrapper
			.find('#title')
			.simulate('change', { target: { value: item.title, name: 'title' } });
		wrapper
			.find('#price')
			.simulate('change', { target: { value: item.price, name: 'price' } });
		wrapper.find('#description').simulate('change', {
			target: { value: item.desc, name: 'desc' },
		});
		expect(wrapper.find('CreateItem').instance().state).toMatchObject({
			title: item.title,
			price: item.price,
			desc: item.desc,
			image: '',
			largeImg: '',
		});
		Router.router = { push: jest.fn() };
		wrapper.find('form').simulate('submit');
		await wait(100);
		expect(Router.router.push).toHaveBeenCalledWith({
			pathname: '/item',
			query: { id: '12345' },
		});
	});
});
