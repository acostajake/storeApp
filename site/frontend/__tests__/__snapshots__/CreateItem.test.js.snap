// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem/> matches snapshot 1`] = `
<form
  className="Form-opr7l5-0 gCHedR"
  data-test="form"
  onSubmit={[Function]}
>
  <h2>
    Sell Item
  </h2>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="image"
    >
      Add Image
      <input
        id="file"
        name="image"
        onChange={[Function]}
        placeholder="Add an image"
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        disabled={false}
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        id="description"
        name="desc"
        onChange={[Function]}
        placeholder="Description"
        required={true}
        type="text"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
